# 4.8.2 Grouping Data & Aggregating Variables

## Table of Contents
#### 1. Importing Libraries and Data Files
#### 2. Aggregated mean calculation of the order_number column grouped by department_id
#### 3. How do the results for the entire dataframe differ from those of the subset?
#### 4. Creating a loyalty flag
#### 5. Spending habits by loyalty category
#### 6. Creating a spending flag for each user
#### 7. Creating an order frequency flag for each user

## 1. Importing Libraries and Data Files

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

### Importing Data Files

In [2]:
# Creating main path
path = r'/Users/helenarobalinho/Desktop/Data Analytics/2. Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/03-2026 Instacart Basket Analysis'

In [3]:
# Importing orders_products_merged_derived.pkl file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_merged_derived.pkl'))

In [4]:
# Checking the Shape of the Dataframe
ords_prods_merge.shape

(32404859, 19)

## 2. Aggregated mean calculation of the order_number column grouped by department_id
### (entire dataframe)

In [5]:
# 1. Grouping by 'department_id' 
# 2. Applying the agg() function to each group to obtain the mean values for the "order_number" column

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

,order_number
,mean
department_id,
1,15.457838
2,17.277920
3,17.170395
4,17.811403
5,15.215751
6,16.439806
7,17.225802
8,15.340650


## 3. How do the results for the entire dataframe differ from those of the subset?

The subset averages differ from the full dataset averages because the subset is a sample and may not fully represent the distribution of each department. Small differences are expected due to natural sampling variability, but larger deviations (like in department 10 where the mean difference is -1.515 and department 21 where the mean difference is +2.633) suggest that the subset may over/under represent certain values. This indicates that analyses based solely on the subset may be biased compared to calculations using the complete data.
The subset was practical and efficient for the exploratory analysis, now the full dataframe is needed for the final accurate results.

## 4. Creating a loyalty flag

Criteria:
- If the maximum orders the user has made is over 40,
  then the customer will be labeled a "Loyal customer."
- If the maximum orders the user has made is over 10 but less than or equal to 40, 
then the customer will be labeled a "Regular customer."
- If the maximum orders the user has made is less than or equal to 10, 
then the customer will be labeled a "New customer."

In [6]:
# Aggregating data with the transform() function
ords_prods_merge ['max_order'] = ords_prods_merge.groupby(['user_id']) ['order_number'].transform(np.max)

/var/folders/ff/tlfsdsyn5jnfl92hv4dxd79m0000gn/T/ipykernel_1696/2350222263.py:2: FutureWarning: The provided callable <function max at 0x108e85800> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge ['max_order'] = ords_prods_merge.groupby(['user_id']) ['order_number'].transform(np.max)


In [7]:
# Checking the output
ords_prods_merge.head(15)

,product_id,product_name,aisle_id,department_id,prices,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,First_order,add_to_cart_order,reordered,_merge,price_range_loc,busiest_day,busiest_days,busiest_period_of_day,max_order
0,1,Chocolate Sandwich Cookies,61,19,5.8,3139998,138,28,6,11,3.0,False,5,0,both,Mid-range product,Regularly busy,Regular days,Most orders,32
1,1,Chocolate Sandwich Cookies,61,19,5.8,1977647,138,30,6,17,20.0,False,1,1,both,Mid-range product,Regularly busy,Regular days,Average orders,32
2,1,Chocolate Sandwich Cookies,61,19,5.8,389851,709,2,0,21,6.0,False,20,0,both,Mid-range product,Busiest day,Busiest days,Average orders,5
3,1,Chocolate Sandwich Cookies,61,19,5.8,652770,764,1,3,13,NaN,True,10,0,both,Mid-range product,Regularly busy,Slowest days,Most orders,3
4,1,Chocolate Sandwich Cookies,61,19,5.8,1813452,764,3,4,17,9.0,False,11,1,both,Mid-range product,Least busy,Slowest days,Average orders,3
5,1,Chocolate Sandwich Cookies,61,19,5.8,1701441,777,16,1,7,26.0,False,7,0,both,Mid-range product,Regularly busy,Busiest days,Average orders,26
6,1,Chocolate Sandwich Cookies,61,19,5.8,1871483,825,3,2,14,30.0,False,2,0,both,Mid-range product,Regularly busy,Regular days,Most orders,9
7,1,Chocolate Sandwich Cookies,61,19,5.8,1290456,910,12,3,10,30.0,False,1,0,both,Mid-range product,Regularly busy,Slowest days,Most orders,12
8,1,Chocolate Sandwich Cookies,61,19,5.8,369558,1052,10,1,20,19.0,False,1,0,both,Mid-range product,Regularly busy,Busiest days,Average orders,20
9,1,Chocolate Sandwich Cookies,61,19,5.8,589712,1052,15,1,12,15.0,False,2,1,both,Mid-range product,Regularly busy,Busiest days,Most orders,20


The column max_order is created. It is the max number of orders made by each user ID.

In [8]:
# Creating a 'loyalty' flag using the loc() function
ords_prods_merge.loc [ords_prods_merge ['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc [(ords_prods_merge ['max_order'] <= 40) & (ords_prods_merge ['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc [ords_prods_merge ['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [9]:
# Printing the frequency of the loyalty flags:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [10]:
# Checking the updated output
ords_prods_merge.head(15)

,product_id,product_name,aisle_id,department_id,prices,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,...,First_order,add_to_cart_order,reordered,_merge,price_range_loc,busiest_day,busiest_days,busiest_period_of_day,max_order,loyalty_flag
0,1,Chocolate Sandwich Cookies,61,19,5.8,3139998,138,28,6,11,...,False,5,0,both,Mid-range product,Regularly busy,Regular days,Most orders,32,Regular customer
1,1,Chocolate Sandwich Cookies,61,19,5.8,1977647,138,30,6,17,...,False,1,1,both,Mid-range product,Regularly busy,Regular days,Average orders,32,Regular customer
2,1,Chocolate Sandwich Cookies,61,19,5.8,389851,709,2,0,21,...,False,20,0,both,Mid-range product,Busiest day,Busiest days,Average orders,5,New customer
3,1,Chocolate Sandwich Cookies,61,19,5.8,652770,764,1,3,13,...,True,10,0,both,Mid-range product,Regularly busy,Slowest days,Most orders,3,New customer
4,1,Chocolate Sandwich Cookies,61,19,5.8,1813452,764,3,4,17,...,False,11,1,both,Mid-range product,Least busy,Slowest days,Average orders,3,New customer
5,1,Chocolate Sandwich Cookies,61,19,5.8,1701441,777,16,1,7,...,False,7,0,both,Mid-range product,Regularly busy,Busiest days,Average orders,26,Regular customer
6,1,Chocolate Sandwich Cookies,61,19,5.8,1871483,825,3,2,14,...,False,2,0,both,Mid-range product,Regularly busy,Regular days,Most orders,9,New customer
7,1,Chocolate Sandwich Cookies,61,19,5.8,1290456,910,12,3,10,...,False,1,0,both,Mid-range product,Regularly busy,Slowest days,Most orders,12,Regular customer
8,1,Chocolate Sandwich Cookies,61,19,5.8,369558,1052,10,1,20,...,False,1,0,both,Mid-range product,Regularly busy,Busiest days,Average orders,20,Regular customer
9,1,Chocolate Sandwich Cookies,61,19,5.8,589712,1052,15,1,12,...,False,2,1,both,Mid-range product,Regularly busy,Busiest days,Most orders,20,Regular customer


## 5. Spending habits by loyalty category

In [11]:
# Determining basic statistics of spending habits by loyalty category:
# Loyal custumer, Regular customer, New customer

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers tend to purchase lower-cost items, with an average price of 10.39, while new customers spend more, with a mean cost of 13.29. Regular customers fall in the middle, averaging 12.50 per item. 
All groups share a minimum price of 1.0, but the extremely high maximum cost of 99,999 suggests possible outliers or data errors.

## 6. Creating a spending flag for each user

Criteria:
- If the mean of the prices of products purchased by a user is lower than 10,
  then flag them as a "Low spender."
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a "High spender."

In [12]:
# Aggregating data with the transform() function
ords_prods_merge ['avg_product_price'] = ords_prods_merge.groupby(['user_id']) ['prices'].transform(np.mean)

/var/folders/ff/tlfsdsyn5jnfl92hv4dxd79m0000gn/T/ipykernel_1696/2747692012.py:2: FutureWarning: The provided callable <function mean at 0x108e86200> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge ['avg_product_price'] = ords_prods_merge.groupby(['user_id']) ['prices'].transform(np.mean)


In [13]:
# Checking relevant columns 
ords_prods_merge[['user_id', 'avg_product_price']].head(15)

,user_id,avg_product_price
0,138,6.935811
1,138,6.935811
2,709,7.930208
3,764,4.972414
4,764,4.972414
5,777,6.935398
6,825,5.957576
7,910,6.680000
8,1052,7.162500
9,1052,7.162500


In [14]:
# Creating a spending flag using the loc() function
ords_prods_merge.loc [ords_prods_merge ['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc [ords_prods_merge ['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [15]:
# Printing the frequency of the spending flags:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [16]:
# Checking relevant columns 
ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

,user_id,avg_product_price,spending_flag
0,138,6.935811,Low spender
1,138,6.935811,Low spender
2,709,7.930208,Low spender
3,764,4.972414,Low spender
4,764,4.972414,Low spender
5,777,6.935398,Low spender
6,825,5.957576,Low spender
7,910,6.680000,Low spender
8,1052,7.162500,Low spender
9,1052,7.162500,Low spender


In [17]:
# The team now can target different types of spenders in their marketing campaigns

## 7. Creating an order frequency flag for each user

Criteria:

- If the median of "days_since_prior_order" is higher than 20, then the customer should be labeled a "Non-frequent customer."

- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a "Regular customer."

- If the median is lower than or equal to 10, then the customer should be labeled a "Frequent customer."

In [18]:
# Aggregating data with the transform() function
ords_prods_merge ['median_days_between_orders'] = ords_prods_merge.groupby(['user_id']) ['days_since_prior_order'].transform(np.median)

/var/folders/ff/tlfsdsyn5jnfl92hv4dxd79m0000gn/T/ipykernel_1696/1392897766.py:2: FutureWarning: The provided callable <function median at 0x108fbbc40> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge ['median_days_between_orders'] = ords_prods_merge.groupby(['user_id']) ['days_since_prior_order'].transform(np.median)


In [19]:
# Checking relevant columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

,user_id,days_since_prior_order,median_days_between_orders
0,138,3.0,8.0
1,138,20.0,8.0
2,709,6.0,8.0
3,764,NaN,9.0
4,764,9.0,9.0
5,777,26.0,11.0
6,825,30.0,20.0
7,910,30.0,6.0
8,1052,19.0,10.0
9,1052,15.0,10.0


In [20]:
# # Creating a order frequency flag using the loc() function
ords_prods_merge.loc [ords_prods_merge ['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc [(ords_prods_merge ['median_days_between_orders'] > 10) & (ords_prods_merge ['median_days_between_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc [ords_prods_merge ['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [21]:
# Printing the frequency of the order frequency flags:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [22]:
# Checking relevant columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

,user_id,days_since_prior_order,median_days_between_orders,order_frequency_flag
0,138,3.0,8.0,Frequent customer
1,138,20.0,8.0,Frequent customer
2,709,6.0,8.0,Frequent customer
3,764,NaN,9.0,Frequent customer
4,764,9.0,9.0,Frequent customer
5,777,26.0,11.0,Regular customer
6,825,30.0,20.0,Regular customer
7,910,30.0,6.0,Frequent customer
8,1052,19.0,10.0,Frequent customer
9,1052,15.0,10.0,Frequent customer


In [23]:
# The Instacart team whave now information of frequent versus non-frequent customers.
# Different types of spenders can be targeted in their marketing campaigns.

## Export dataframe in pickle format to the "Prepared Data" folder

In [24]:
# Exporting dataframe
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))

In [25]:
ords_prods_merge.shape

(32404859, 25)